Remove KSF parameters and mention config updates

chris-wood opened the pull request:
Closes #400
Closes #399 
Closes #398 
Closes #394 
Closes #392

daxpedda commented:
This doesn't address the issue in #400 and only partially addresses #398 (by removing `params`):
the question of what size of salt to use. Specifically to either use none if allowed by the spec or to use all zeroes.

chris-wood commented:
The point of this change is that the choice of KSF parameters is _entirely up the application_. It is not something that OPAQUE specifies. Thus, I believe it addresses all the issues referenced.

daxpedda commented:
I agree with that approach, imo OPAQUE should not deal with KSF parameters.

But there is a problem we might want to address one way or another: currently there is no consensus on the salt issue and it will eventually lead to incompatible implementations (e.g. https://github.com/facebook/opaque-ke/pull/275).

It would help to add a note just saying what to do here, which is actually still unclear. If I had nothing else but the spec to go by, I would not use a salt at all, which no implementation I know of does.

If it's still decided that we don't want to mention a salt at all in the spec, I would suggest an alternative solution. Currently Argon2 is the recommended algorithm, like the other recommended algorithms (P-*, Ristretto255), maybe adding test vectors would help. Which was requested in https://github.com/cfrg/draft-irtf-cfrg-opaque/issues/399, I assume the idea was rejected because we don't want to make it look like specific KSF parameters are suggested?

daxpedda commented:
In `opaque-ke` the output size is determined by the input size, which is the output size of the hash function used in the OPRF. So `Nh`, according to the spec.

I was wondering how that happened, but I found other implementations which did the same: [stef's implementation](https://github.com/stef/libopaque/blob/3ff26cd7fd5fa618361c33cf0c0c7e563a252a71/src/opaque.c#L208-L212), for example, did this too.

This is another point implementations have silently agreed on without the specification. This problem seems to get bigger as we go :laughing:.

I have to say that personally I don't have a stake in this, my use-cases don't require compatibility between multiple implementations, I'm using the same on both sides. Maybe somebody else with an actual stake in this will have more to say.

daxpedda commented on draft-irtf-cfrg-opaque.md:
```diff
@@ -1677,8 +1680,9 @@ parameters that are needed to prevent cross-protocol or downgrade attacks.
 - ristretto255-SHA512, HKDF-SHA-512, HMAC-SHA-512, SHA-512,
-    Argon2id(t=1, p=4, m=2^21), ristretto255
-- P256-SHA256, HKDF-SHA-256, HMAC-SHA-256, SHA-256, Argon2id(t=1, p=4, m=2^21), P-256
+    Argon2id(S = zeroes(16), p = 4, T = 32, m = 2^21, t = 1, v = 0x13, K = nil, X = "RFCXXXX", y = 2), ristretto255
+- P256-SHA256, HKDF-SHA-256, HMAC-SHA-256, SHA-256,
+    Argon2id(S = zeroes(16), p = 4, T = 32, m = 2^21, t = 1, v = 0x13, K = nil, X = "RFCXXXX", y = 2), P-256
```
That would add another point: associated data.

chris-wood replied:
X is the associated data here.

daxpedda replied:
Yes, what I meant to say is that this would another parameter that is not specified by the spec which will have to be silently agreed on to make implementations compatible.

Basically, until now associated data was left empty (I didn't check more implementations, but kevinlewi's and stef's is empty), now, implementations will probably use `RFCXXXX` as their associated data for all KSF's, even though the spec doesn't say you should do that.

chris-wood commented:
To hopefully avoid further nit picks, I just made the output length be `Nh` and set associated data to `nil`.

armfazh commented on draft-irtf-cfrg-opaque.md:
```diff
@@ -2275,7 +2292,9 @@ outputs computed during the authentication of an unknown or unregistered user. N
 includes the (OPRF, Hash, KSF, KDF, MAC, Group, Context) tuple, where the Group
 matches that which is used in the OPRF. These test vectors were generated using
-draft-10 of {{OPRF}}.
+draft-21 of {{OPRF}}. The KSF used for each test vector is the identity function
```
May I suggest to use argon2id but with short `m`, for example instead `m = 2^21` set `m=8`.

kevinlewi reviewed (approved):
Sorry for the delays in getting to this review, but overall looks good to me!

Although I am probably just reiterating @chris-wood's points: I think that aligning implementations on the usage of OPAQUE with specific key stretching functions is outside of the scope of this spec.

This applies to mentioning anything about KSF salt length (for example), which some applications that use OPAQUE may want to align on... but I see as still being outside of the spec scope.
